src/go/guestagent/pkg/containerd/events_linux.go

```diff
@@ -311,63 +327,51 @@ func execIptablesRules(portMappings nat.PortMap, containerID, namespace, pid str
 // We enter the following rule after the existing rule:
 // DNAT       tcp  --  anywhere             anywhere             tcp dpt:9119 to:10.4.0.22:80.
-func createLoopbackIPtablesRules(containerID, namespace, pid, port, destinationPort string) error {
-	// read the container's cni config to extract the network name
-	nsenterNetworkConfCmd := exec.Command("nsenter", "-t", pid, "-n", "cat", "/etc/cni/net.d/nerdctl-bridge.conflist")
-	output, err := nsenterNetworkConfCmd.CombinedOutput()
-	if err != nil {
-		return err
-	}
-
-	log.Debugf("read the following network conflist for containerID: %s config: %s", containerID, string(output))
-
-	var networkConfig cniNetworkConfig
-	err = json.Unmarshal(output, &networkConfig)
-	if err != nil {
-		return err
-	}
-
+func createLoopbackIPtablesRules(ctx context.Context, networks []string, containerID, namespace, pid, port, destinationPort string) error {
 	eth0IP, err := extractIPAddress(pid)
 	if err != nil {
 		return err
 	}
 
 	log.Debugf("found the ip address: %s for containerID: %s", eth0IP, containerID)
-
-	// create the corresponding chain name, e.g CNI-DN-xxxxxx
-	// (where xxxxxx is a function of the ContainerID and network name)
-	// https://www.cni.dev/plugins/current/meta/portmap/#dnat
-	// https://github.com/containernetworking/plugins/blob/2b097c5a62dacfd2f9ea4268faa4fc04bd5343f5/pkg/utils/utils.go#L39
 	cID := fmt.Sprintf("%s-%s", namespace, containerID)
-	chainName := mustFormatHashWithPrefix(maxChainLength, chainPrefix+"DN-", networkConfig.Name+cID)
-
-	log.Debugf("determined iptables chain name: %s for containerID: %s", chainName, containerID)
-
-	// Instead of updating the existing rule we insert the overriding rule below the previous one
-	// e.g rule can be:
-	// iptables -t nat -A CNI-DN-xxxxxx -p tcp -d 0.0.0.0/0 -j DNAT --dport 9119 --to-destination 10.4.0.10:80
-	iptableCmd := exec.Command("iptables",
-		"--table", "nat",
-		"--append", chainName,
-		"--protocol", "tcp",
-		"--destination", "0.0.0.0/0",
-		"--jump", "DNAT",
-		"--dport", destinationPort,
-		"--to-destination", fmt.Sprintf("%s:%s", eth0IP, port))
-
-	return iptableCmd.Run()
-}
 
-func (e *EventMonitor) createPortMapping(ctx context.Context, namespace, containerID string) (nat.PortMap, error) {
-	container, err := e.containerdClient.ContainerService().Get(
-		containerdNamespace.WithNamespace(ctx, namespace), containerID)
-	if err != nil {
-		return nil, err
+	var allErrs []error
+
+	// Run the rule per network
+	for _, network := range networks {
+		chainName := utils.MustFormatChainNameWithPrefix(network, cID, "DN-")
+
+		// Instead of modifying the existing rule, we add a new rule that overrides the previous one.
+		// The new rule is appended below the existing rule in the chain, ensuring that traffic is correctly
+		// routed to the specified destination. An example of the rule looks like:
+		//   iptables -t nat -A CNI-DN-xxxxxx -p tcp -d 0.0.0.0/0 -j DNAT --dport 9119 --to-destination 10.4.0.10:80
```

Comment:
For reference, this works because (after inserting the rule) the table looks like:

```
Chain CNI-DN-51ff1a317aa0f37fe3b51 (1 references)
target     prot opt source               destination
CNI-HOSTPORT-SETMARK  tcp  --  10.4.1.0/24          localhost            tcp dpt:8080
CNI-HOSTPORT-SETMARK  tcp  --  localhost            localhost            tcp dpt:8080
DNAT       tcp  --  anywhere             localhost            tcp dpt:8080 to:10.4.1.2:80
DNAT       tcp  --  anywhere             anywhere             tcp dpt:8080 to:10.4.1.2:80
```
The previous rule is anywhere → localhost, so the traffic from anywhere → anywhere would otherwise fall off the end of the chain.  (The comment was worded like the traffic would have otherwise hit a different rule first, which got me confused as to why it would work.)